### PR TITLE
#161 :sparkles: migration dev-dependencies to group.dev.dependencies …

### PR DESCRIPTION
**Pull request by d-ryan-ashcraft**
…automatically

### CODE REVIEW

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/AbstractTomlGroupMigration.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
+package org.technologybrewery.habushu.migration;
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/AbstractTomlGroupMigration.java:1]: I (Informational): refactored existing group migration to an abstract class to make it easy to add multiple

File: habushu-maven-plugin/src/test/resources/migration/dev-dependencies/with-no-dev-dependenies-group.toml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[tool.poetry]
```
Comment (Cho-William) [inline habushu-maven-plugin/src/test/resources/migration/dev-dependencies/with-no-dev-dependenies-group.toml]: this file's spelling seems to be missing a "c"

File: habushu-maven-plugin/src/main/resources/migrations.json [hunk 3]
```diff
@@ -1,26 +1,41 @@
...
+        "fileSets": [
+          {
+            "includes": [
+              "pyproject.toml"
+            ]
+          }
+        ]
+      },
+      {
+        "name": "dev-dependencies-preferred-poetry-approach-group-migration",
```
Comment (Cho-William) [inline habushu-maven-plugin/src/main/resources/migrations.json:29]: A: I personally think this reads better, but no worries if we think the original naming is better 

```suggestion
        "name": "preferred-poetry-approach-dev-dependencies-group-migration",
```

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    @Then("{int} pyproject.toml dependencies exist in the tool.poetry.dev-dependencies group")
+    public void pyproject_toml_dependencies_exist_in_the_tool_poetry_dev_dependencies_group(Integer expectedDevDashDependenciesGroupDependencies) {
+        verifyExecutionOccurred();
+        try (FileConfig tomlFileConfig = FileConfig.of(pyProjectToml)) {
+            tomlFileConfig.load();
+
+            Optional<Config> toolPoetryDevDependencies = tomlFileConfig.getOptional("tool.poetry.dev-dependencies");
+            int numberOfLegacyDependencies = getNumberOfChildren(toolPoetryDevDependencies);
+            assertEquals(expectedDevDashDependenciesGroupDependencies, numberOfLegacyDependencies,
+                    expectedDevDashDependenciesGroupDependencies + " dev-dependencies group should not exist!");
```
Comment (Cho-William) [inline habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java]: A: The string message seems a bit off? Given that we are expecting `expectedDevDashDependenciesGroupDependencies` to be zero, shouldn't this read as:
```suggestion
                    expectedDevDashDependenciesGroupDependencies + " dev-dependencies should exist!");
```

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    @Then("{int} pyproject.toml dependencies exist in the tool.poetry.group.dev.dependencies group")
+    public void pyproject_toml_dependencies_exist_in_the_tool_poetry_group_dev_dependencies_group(Integer expectedGroupDevDependencies) {
+        verifyExecutionOccurred();
+        try (FileConfig tomlFileConfig = FileConfig.of(pyProjectToml)) {
+            tomlFileConfig.load();
+
+            Optional<Config> toolPoetryDependencies = tomlFileConfig.getOptional(TomlUtils.TOOL_POETRY_DEV_DEPENDENCIES);
+            int numberOfGroupDevDependencies = getNumberOfChildren(toolPoetryDependencies);
+            assertEquals(expectedGroupDevDependencies, numberOfGroupDevDependencies,
+                    expectedGroupDevDependencies + " group.dev.depependenies should remain in main group!");
```
Comment (Cho-William) [inline habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java:64]: See above comment.

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    protected int getNumberOfChildren(Optional<Config> tomlElement) {
+        int numberOfLegacyDependencies = 0;
+        if (tomlElement.isPresent()) {
+            Config foundDependencies = tomlElement.get();
+            Map<String, Object> dependencyMap = foundDependencies.valueMap();
+
+            for (Map.Entry<String, Object> dependency : dependencyMap.entrySet()) {
+                Object packageRhs = dependency.getValue();
+                numberOfLegacyDependencies++;
```
Comment (Cho-William) [inline habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java]: Q: is there a reason why we can't just get the `.size()` of the `entrySet`?

**Review by Cho-William — APPROVED** [review]
All feedback was very minor, LGTM!

**Review by epwilkins — APPROVED** [review]

File: habushu-maven-plugin/src/test/resources/migration/dev-dependencies/with-no-dev-dependenies-group.toml (hunk 2 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/test/resources/migration/dev-dependencies/with-no-dev-dependenies-group.toml]: ✅ Good catch - fixed!

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java (hunk 4 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java]: I borrowed this from another class - I think intent is that this check is always looking for the passed number to not be there.  Updated to ` is an unexpected number of dev-dependencies`.  it happens to be zero in this case, but we could add more tests and that would no longer be the case.

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java (hunk 5 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java:64]: I think this one is good as is

File: habushu-maven-plugin/src/main/resources/migrations.json (hunk 3 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/main/resources/migrations.json:29]: TomAto-tOmato - gonna leave it the way it is

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java (hunk 5 above)
Comment (Cho-William) [reply habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java:64]: agreed, was reading too fast

File: habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java (hunk 6 above)
Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/test/java/org/technologybrewery/habushu/migration/DevDependenciesGroupMigrationSteps.java]: ✅ Good catch - refactored this from where it was using the values, but didn't simplify after doing that.  FIXED.

**Pull request merged**